packages/phalanx/example/FiniteElementAssembly_NoTags/evaluators/GatherSolution_Def.hpp

```diff
@@ -0,0 +1,204 @@
+// @HEADER
+// ************************************************************************
+//
+//        Phalanx: A Partial Differential Equation Field Evaluation 
+//       Kernel for Flexible Management of Complex Dependency Chains
+//                    Copyright 2008 Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roger Pawlowski (rppawlo@sandia.gov), Sandia
+// National Laboratories.
+//
+// ************************************************************************
+// @HEADER
+
+
+#include "Teuchos_Assert.hpp"
+#include "Phalanx_DataLayout.hpp"
+
+// **********************************************************************
+// Specialization: Residual
+// **********************************************************************
+
+template<typename Traits>
+GatherSolution<PHX::MyTraits::Residual, Traits>::
+GatherSolution(const std::string& field_name,
+               const Teuchos::RCP<const PHX::DataLayout>& layout,
+               const int& in_num_equations,
+               const int& in_field_index,
+               const Kokkos::View<double*,PHX::Device>& in_x) :
+  num_equations(in_num_equations),
+  field_index(in_field_index),
+  x(in_x)
+{
+  PHX::Tag<ScalarT> field_tag(field_name,layout);
+  this->addEvaluatedField(field_tag,field);
```

Comment:
For now yes. Here's the issue - if we want to mix MDField, Field and Views, then the registered Views must conform to the declaration that the MDField and Field objects expect so that they can dynamic_cast them from the any object. I had originally written the View acceptors to any arbitrary view and then encountered casting failures in testing. If we need to do that in the future, I think I can get it to work but it was unsafe and would take changes in many places. So for now, by forcing the View object to use the PHX::Device, the signature is consistent and it works with all wrappers. If we want to loosen the criteria we can do it in the future.